### PR TITLE
Adding an experimentally measured temporal pulse profile

### LINKED ISSUE #60: 1D Complex Temporal Profile, e.g., FROG data
This is a longitudinal profile: get the complex longitudinal envelope (both amplitude **and** phase) from FROG measurements.

### PR DESCRIPTION
**Pull request by rob-shalloo**
Incorporates experimentally measured pulse profiles in either the temporal or spectral domain.

Closes #60.

### CODE REVIEW

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    ----------
+    data : structure
+        The data structure comprises several items indexed by
+        a series of keys
+
+        datatype : string
+            The domain in which the data has been passed. Options
+            are 'spectral' and 'temporal'
+
+        abscissa : ndarrays of floats
```
Comment (MaxThevenet) [inline lasy/profiles/longitudinal/longitudinal_profile_from_data.py]: Could we us "axis" instead? This is the term used in other places.

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+            When datatype is 'spectral' abscissa is wavelength in
+            meters
+            When datatype is 'temporal' abscissa is time in seconds
+
+        intensity : ndarrays of floats
+            The vertical axis of the pulse duration measurement
+            When datatype is 'spectral' intensity is spectral
+            intensity. This will be noramlized
+            When datatype is 'temporal' intensity is temporal
+            intensity. This will be normalized
```
Comment (MaxThevenet) [inline lasy/profiles/longitudinal/longitudinal_profile_from_data.py]:
```suggestion
            The vertical axis of the pulse duration measurement.
            Spectral (resp. temporal) intensity when datatype is 'spectral' (resp 'temporal').
```
Fixing a few typos

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+        phase : ndarrays of floats
```
Comment (MaxThevenet) [inline lasy/profiles/longitudinal/longitudinal_profile_from_data.py]:
```suggestion
        phase : ndarray of floats
```

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            If provided, this phase will be added to the pulse.
+            When datatype is 'spectral' phase is spectral
+            phase.
+            When datatype is 'temporal' phase is temporal
+            phase
```
Comment (MaxThevenet) [inline lasy/profiles/longitudinal/longitudinal_profile_from_data.py]: Any reason to cut those lines? <80 is restrictive enough.
```suggestion
            When datatype is 'spectral' phase is spectral phase.
            When datatype is 'temporal' phase is temporal phase.
```

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+        dt : float
+            Only required when datatype is 'spectral'. In this
+            case this defines the user requested resolution in
+            the conversrion from the spectral to the temporal
```
Comment (MaxThevenet) [inline lasy/profiles/longitudinal/longitudinal_profile_from_data.py]:
```suggestion
            the conversion from the spectral to the temporal
```

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            domain.
+
+        wavelength : float
+            Only required when datatype is 'temporal'. Then,
+            this is the central wavelength of the pulse
+
+    lo, hi : floats (seconds)
+        Lower and higher end of the required domain of the data.
```
Comment (MaxThevenet) [inline lasy/profiles/longitudinal/longitudinal_profile_from_data.py]:
```suggestion
        Lower and higher ends of the required domain of the data.
```

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+            spectral_phase = data["phase"]
+            dt = data["dt"]
+            cwl = np.sum(spectral_intensity * wavelength) / np.sum(spectral_intensity)
+            cfreq = c / cwl
+            # Determine required sampling frequency for desired dt
+            sample_freq = 1 / dt
+            # Determine number of points in temporal domain
+            indx = np.argmin(np.abs(wavelength - cwl))
+            dfreq = np.abs(c / wavelength[indx] - c / wavelength[indx + 1])
+            N = int((sample_freq) / dfreq)
```
Comment (MaxThevenet) [inline lasy/profiles/longitudinal/longitudinal_profile_from_data.py]: Also, the user could just specify the number of points in temporal domain, as an option. Would that be easy?
```suggestion
            N = int(sample_freq / dfreq)
```

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        self.temporal_intensity = temporal_intensity[tIndLo:tIndHi]
+        self.temporal_phase = temporal_phase[tIndLo:tIndHi]
+
+    def evaluate(self, t):
+        """
+        Returns the longitudinal field envelope
+
+        Parameters
+        ----------
+        t: ndarrays of floats
```
Comment (MaxThevenet) [inline lasy/profiles/longitudinal/longitudinal_profile_from_data.py]:
```suggestion
        t : ndarray of floats
```

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+            Define points on which to evaluate the envelope
+
+        Returns:
+        --------
```
Comment (MaxThevenet) [inline lasy/profiles/longitudinal/longitudinal_profile_from_data.py]:
```suggestion
        Returns
        -------
```

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+        envelope: ndarray of complex numbers
```
Comment (MaxThevenet) [inline lasy/profiles/longitudinal/longitudinal_profile_from_data.py]:
```suggestion
        envelope : ndarray of complex numbers
```

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+            # First find central frequency
+            wavelength = data["abscissa"]
+            spectral_intensity = data["intensity"]
+            spectral_phase = data["phase"]
+            dt = data["dt"]
+            cwl = np.sum(spectral_intensity * wavelength) / np.sum(spectral_intensity)
+            cfreq = c / cwl
+            # Determine required sampling frequency for desired dt
+            sample_freq = 1 / dt
+            # Determine number of points in temporal domain
```
Comment (MaxThevenet) [inline lasy/profiles/longitudinal/longitudinal_profile_from_data.py]: Could you add 1 line explaining how you calculate it?

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+                )
+                / dt
+            )
+            time = np.linspace(-dt * N / 2, dt * N / 2 - dt, N)
+
+            # Extract intensity and phase
+            temporal_intensity = np.abs(t_amplitude) ** 2
+            temporal_intensity /= np.max(temporal_intensity)
+            temporal_phase = np.unwrap(-np.angle(t_amplitude))
+            temporal_phase -= temporal_phase[np.argmin(np.abs(time))]
```
Comment (MaxThevenet) [inline lasy/profiles/longitudinal/longitudinal_profile_from_data.py:104]: Should the phase be defined with a `-` (as currently done in these 2 lines) or with a `+` sign? I am not sure what the most natural/usual convention is here, considering that we use `exp(-j*omega0*t)` in the definition of the envelope. Note that if we change it here, the `evaluate`function below also needs to be updated.

**Review by MaxThevenet — COMMENTED** [review]
Thanks for this PR! See comments below. Also, have you been able to test it?

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py (hunk 1 above)
Comment (rob-shalloo) [reply lasy/profiles/longitudinal/longitudinal_profile_from_data.py]: Sure, I believe abscissa is technically the correct name for the horizontal axis of a plot, but axis will certainly make more sense to the general user and make the code less opaque

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py (hunk 11 above)
Comment (rob-shalloo) [reply lasy/profiles/longitudinal/longitudinal_profile_from_data.py]: Done

File: lasy/profiles/longitudinal/longitudinal_profile_from_data.py (hunk 12 above)
Comment (rob-shalloo) [reply lasy/profiles/longitudinal/longitudinal_profile_from_data.py:104]: So it has been tested and the intensity lineout extracted from evaluating the longitudinal profile or from indexing directly into `laser.field.field` correctly reproduces the temporal profile (the correct way around, so think the +/- is correct). However, currently we don't match with reading back in the openPMD output, neither in `xyt` or `rt` but I think this is a future problem

Screenshot 2023-05-18 at 17 22 09

Comment (MaxThevenet) [reply lasy/profiles/longitudinal/longitudinal_profile_from_data.py:104]: Thanks for clarifying!

**Review by MaxThevenet — APPROVED** [review]
Thanks!

**Pull request merged**